src/Shared/BuildEnvironmentHelper.cs

```diff
@@ -432,23 +432,37 @@ public BuildEnvironment(BuildEnvironmentMode mode, string currentMSBuildExePath,
             CurrentMSBuildToolsDirectory = Path.GetDirectoryName(currentMSBuildExePath);
             CurrentMSBuildConfigurationFile = string.Concat(currentMSBuildExePath, ".config");
 
+            // We can't detect an environment, don't try to set paths.
+            if (mode == BuildEnvironmentMode.None)
+                return;
+
             VisualStudioInstallRootDirectory = visualStudioPath;
 
-            if (mode == BuildEnvironmentMode.VisualStudio)
-            {
-                var isAmd64 = FileUtilities.EnsureNoTrailingSlash(CurrentMSBuildToolsDirectory)
-                    .EndsWith("amd64", StringComparison.OrdinalIgnoreCase);
+            // Check to see if our current folder is 'amd64'
+            bool runningInAmd64 = FileUtilities.EnsureNoTrailingSlash(CurrentMSBuildToolsDirectory)
+                .EndsWith("amd64", StringComparison.OrdinalIgnoreCase);
 
-                if (isAmd64)
-                {
-                    MSBuildToolsDirectory32 = FileUtilities.GetFolderAbove(CurrentMSBuildToolsDirectory);
-                    MSBuildToolsDirectory64 = CurrentMSBuildToolsDirectory;
-                }
-                else
-                {
-                    MSBuildToolsDirectory32 = CurrentMSBuildToolsDirectory;
-                    MSBuildToolsDirectory64 = Path.Combine(CurrentMSBuildToolsDirectory, "amd64");
-                }
+            var msBuildExeName = Path.GetFileName(currentMSBuildExePath);
+            var folderAbove = FileUtilities.GetFolderAbove(CurrentMSBuildToolsDirectory);
+
+            // Calculate potential paths to other architecture MSBuild.exe
+            var potentialAmd64FromX86 = FileUtilities.CombinePaths(CurrentMSBuildToolsDirectory, "amd64", msBuildExeName);
+            var potentialX86FromAmd64 = Path.Combine(folderAbove, msBuildExeName);
+
+            // Check for existence of the file. Not nessisary in VS where we always want to always the layout.
+            var existsCheck = new Func<string, bool>(path => mode == BuildEnvironmentMode.VisualStudio || File.Exists(path));
+
+            // Running in amd64 folder and the X86 is valid
+            if (runningInAmd64 && existsCheck(potentialX86FromAmd64))
+            {
+                MSBuildToolsDirectory32 = folderAbove;
+                MSBuildToolsDirectory64 = CurrentMSBuildToolsDirectory;
+            }
+            // Not running in amd64 folder and the amd64 path is valid
+            else if (existsCheck(potentialAmd64FromX86))
```

Comment:
Shouldn't this be `(!runningInAmd64 && existsCheck(potentialAmd64FromX86)`. If this else branch hits, it's because either one of the terms in the previous if condition could be false.